consul/autopilot.go

```diff
@@ -0,0 +1,298 @@
+package consul
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/hashicorp/consul/consul/agent"
+	"github.com/hashicorp/consul/consul/structs"
+	"github.com/hashicorp/raft"
+	"github.com/hashicorp/serf/serf"
+)
+
+// AutopilotPolicy is the interface for the Autopilot mechanism
+type AutopilotPolicy interface {
+	// PromoteNonVoters defines the handling of non-voting servers
+	PromoteNonVoters(*structs.AutopilotConfig) error
+}
+
+func (s *Server) startAutopilot() {
+	s.autopilotShutdownCh = make(chan struct{})
+
+	go s.serverHealthLoop()
+	go s.removeDeadLoop()
+}
+
+func (s *Server) stopAutopilot() {
+	close(s.autopilotShutdownCh)
+}
+
+// serverHealthLoop monitors the health of the servers in the cluster
+func (s *Server) serverHealthLoop() {
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(s.config.ServerHealthInterval)
+	for {
+		select {
+		case <-s.autopilotShutdownCh:
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			serverHealths := make(map[string]*structs.ServerHealth)
+
+			state := s.fsm.State()
+			_, autopilotConf, err := state.AutopilotConfig()
+			if err != nil {
+				s.logger.Printf("[ERR] consul: error retrieving autopilot config: %s", err)
+			}
+
+			// Build an updated map of server healths
+			for _, member := range s.LANMembers() {
+				if member.Status == serf.StatusLeft {
+					continue
+				}
+
+				valid, parts := agent.IsConsulServer(member)
+				if valid {
+					health, err := s.queryServerHealth(member, parts, autopilotConf)
+					if err != nil {
+						s.logger.Printf("[ERR] consul: error fetching server health: %s", err)
+					} else {
+						serverHealths[parts.Addr.String()] = health
+					}
+				}
+			}
+
+			s.autopilotLock.Lock()
+			s.autopilotHealth = serverHealths
```

Comment:
I think we need to move things around a bit so this is cleaner from a start/stop perspective. Here's the scenario I'm worried about:

1. One of the health queries takes a long time because the server dropped off but it's TCP.
2. We cycle leadership so this goroutine gets abandoned and a new set gets created.
3. This one eventually comes back and writes to the server health structure with stale stuff.

One thing to make this simpler could be to put the server health checking in a separate loop that doesn't start and stop (just stops on Consul's main `shutdownCh`). This would have the advantage that a server that becomes leader would have fresh health data and not have to wait, and it would prevent the stale problem. If we did that, I think we could safely do a `sync.WaitGroup` on shutdown to wait for the rest of the goroutines to die off, since they are more deterministic - we can add a timeout to the promote Raft call to bound it.